### ISSUE TITLE
[Bug] Data lost from table edits

### ISSUE THREAD

**Issue by bilderbuchi** [Zettlr/Zettlr#3011]
### Description

It seems that table edits are not always saved to file. This caused data loss (hours of work) twice already! :-(

I'm a new user. I am working on a single md file (so, not in a Zettlr directory) that includes a table (~15x15) with plain text and hyperlinks. So far the experience was ok-ish -- sometimes the (very nice!) table rendering fails and the markdown table plaintext becomes visible, but some combination of clicking outside or switching to another file in Zettlr and back typically brings the rendered view back.

Day before yesterday, I mixed up app focus and inadvertently closed the file with a Ctrl-W. Upon opening again I apparently had **lost** a number of table edits, which was quite annoying. I confirmed that the content was not there with a conventional editor.
I was curious at the time why I did not see a "you have unsaved edits, are you sure to close?" dialog. 
Anyway, I redid my work and from now on made sure to press Ctrl-S from time to time and make sure there is no asterisk beside the file title.

Yesterday, I spent a couple of hours filling the table further. Saved with Ctrl-S. Put the laptop to sleep. 
Today,  Zettlr still showed that content. Then I closed the file in Zettlr. Then I checked the contents in another editor --> All edits lost!!

At this point, I'm quite disappointed and surprised that such a data loss could even occur.
Having been burnt before, I am sure that i saved the file, and there was no asterisk beside the file name. So, it seems that Zettlr considers the table edits saved, but this is not reflected on the file on disk. This could explain the observation above that I did not get a confirmation dialog.

Further assorted points:
* The file also contains one image, which is correctly displayed and is stored at the same location as the file
* Aside of that, there are a couple of lists and section headings.
* I only have the Zettlr tutorial workspace and 2 other md files I started in the Zettlr file manager.

What other details do you need?

### Reproducing

I don't have a MWE repro procedure right now. I am busy redoing a lot of lost work.

### Zettlr Version

Stable (most recent version)

### Specify version

2.1.0

### Your Platform

- [ ] Windows
- [ ] macOS
- [X] Linux

### Architecture

- [X] Intel 64bit
- [ ] ARM 64bit

### Operating System Version

Kubuntu 20.04

### Additional Information

* Zettlr was installed via deb

Some more weird/buggy behaviour observations that may help to pin this down (or might be separate bugs):
* Editing normal text immediately triggers the asterisk. Then, Ctrl-W correctly triggers the confirmation dialog
* Editing normal text, then Ctrl-W, then Cancel does actually unexpectedly save the file! (not sure if the dialog triggers an autosave? autosave is on "after a short while")
* Editing normal text, then middle mouse on the tab/file name triggers the confirmation dialog, but _also_ incorrectly pastes the Linux selection clipboard contents at the cursor location. 
* Editing table cells, even multiple, does not trigger the asterisk. One has to click outside the table for that to register.
* Font style shortcuts like Ctrl-B, Ctrl-I on text in a rendered table cell seem to influence the selected text, but as soon as one clicks outside the cell, the text loses the bold or italic styling again
* Sometimes the (very nice!) table rendering fails and the markdown table plaintext becomes visible, but some combination of clicking outside or switching to another file in Zettlr and back, or closing and reopening the file typically brings the rendered view back.

**Comment by bilderbuchi** [comment]
OK, I have a reproducer for at least some issue (with 2.1.0). 
* Save this file as e.g. `zettlr_3011.md`
```
# Test

| Test | Test | Test | Test |
| ---- | ---- | ---- | ---- |
| 1    | 1    | 1    | 1    |
| 2    | 2    | 2    | 2    |
| 3    | 3    | 3    | 3    |
| 4    | 4    | 4    | 4    |



```
* Open the file in Zettlr
* Observe the rendered table editor
* Modify contents, e.g. put an `a` after all `2` entries
* Observe that there is no asterisk by the filename (that would indicate unsaved state)
* Save the file (Ctrl-S)
* Close the file (Ctrl-W)
* The file closes (no top bar anymore), but you get this weird UI state:
![image](https://user-images.githubusercontent.com/327442/148599351-d85c6a74-ba58-42ed-a8e8-33239552c863.png)
* Open the file `zettlr_3011.md` with another editor, notice that your edits were in fact **not** saved.

**Comment by bilderbuchi** [comment]
I just confirmed that the reproducer works identically on Zettlr 2.1.1.

**Comment by nathanlesage** [comment]
The main problem is that the TableEditor is still kind of hacky. It's quite the complex component, but the main problem is that while you edit things, we cannot write those changes you do in the table back every once in a while since writing that back to the editor will un-render the table so that it has to be rendered again. This means in order to have those changes actually applied to the underlying document you must first un-focus the table (i.e. click anywhere into the document). This will trigger the table to rewrite any changes back to the document and that is when the asterisk appears.

I know that this is definitely not a smooth experience, but I've been really busy with other changes that I couldn't get to that until now, even though I'm aware of this since a long time now :( 

So until I can get back to this, I would recommend to, from time to time, click somewhere outside the table, save the file, and continue. This should prevent any data loss.

**Comment by bilderbuchi** [comment]
Thank you for the analysis!

> This means in order to have those changes actually applied to the underlying document you must first un-focus the table (i.e. click anywhere into the document).

Can you trigger such an un-focus/re-focus/rerender cycle when going from one cell to another? I think at that point it can be expected that the state of edits is "complete".

In relation to that, I just played around with clicking outside the table editor and into it again, and the behaviour did not seem consistent to me. Is it expected that raw markdown is displayed _every time_ one clicks outside the table (not my experience - sometimes the table stayed)? 
After that, what is the recommended way to obtain a rendered table, again? Is it really "clicking inside then outside again", which matches my observations?

>  even though I'm aware of this since a long time now :(

I unfortunately missed the box cautioning about this limitation in the docs because the rest of the table editor was intuitive enough that I did not feel the need to read its documentation -- so, good job, I guess :grin: 

> click somewhere outside the table, save the file, and continue.

thank you, this seems like a feasible workaround for now until this is resolved.

**Comment by bilderbuchi** [comment]
> This should prevent any data loss.

As mentioned above, considering that I had to re-obtain table rendering multiple times, I'm really not sure that I have never clicked outside the table in several hours of editing this table. So, I'm not sure if this alone caused the observed data loss.

TBH, I was a bit shocked that such a severe data loss could even occur, so when using Zettlr to build up a notes collection I'm of half a mind to wrap a git repo or somesuch around to confirm if I can trust the editor, which is not a good position to be in. :-/
I've also noticed in the issue tracker that a couple of other data loss bugs seem to be present -- maybe it makes sense to label/triage those accordingly.

**Comment by nathanlesage** [comment]
@bilderbuchi I fully understand your concerns, and I agree that the TableEditor is definitely still wonky. I do plan on improving that behavior, but please understand that it's a super complicated part of the app and I have to do that in my free time. However, I'm slowly coming to a point where I can get back to the TableEditor.

In the meantime, the suggested workaround should be good to go. Yes, it's definitely not the best way, but it will for sure prevent data loss.

With regard to the suggestion to write the contents back into the editor when switching cells, this could be an idea. It might take a while to implement that, however, because it requires a pretty good timing so users don't really see this "invisible" switching 🙈

**Comment by bilderbuchi** [comment]
Sure, please take things at your pace (I know the situation well :sweat_smile: )! I can live with bugs and wonkiness, as long as it does not destroy a significant amount of work...

Could be that users operating markdown editors (which IME often feature "previews", live-generated or otherwise) are actually used to a little invisible switching/waiting ^^

**Comment by nathanlesage** [comment]
>  are actually used to a little invisible switching/waiting ^^

Yes, definitely. I think it's perfectly fine for a very minor visual glitch while everything is persisted, but it is actually quite hard to do so. But I'll figure out a way!

**Comment by nathanlesage** [comment]
Small update: I took my time to migrate the TableEditor to TypeScript now, and I found quite a few things this way that went unnoticed before.

Additionally, I ensured that Setext headings level 2 without a following newline are no longer detected as tables. That is, if you accidentally write

```
This is a heading
-----------------
Immediately a paragraph
```

this will no longer be rendered as a table. (However, this makes this not a valid Setext heading …)
